File: solutions/ocaml/phone-number/1/phone_number.ml
open Base

(* ---- Validation helpers ---- *)
let valid_characters = "0123456789 ()+.-"

let contains_letters input =
  String.exists input ~f:Char.is_alpha

let is_invalid_character character =
  not (String.contains valid_characters character)

let contains_invalid_punctuation input =
  String.exists input ~f:is_invalid_character

(* ---- Cleaning ---- *)
let extract_digits input =
  input
  |> String.to_list
  |> List.filter ~f:Char.is_digit
  |> String.of_char_list

(* ---- Length check ---- *)
let drop_country_code digits =
  String.sub digits ~pos:1 ~len:10

let validate_eleven_digit_number digits =
  let first_digit = String.get digits 0 in
  if Char.equal first_digit '1'
  then Ok (drop_country_code digits)
  else Error "11 digits must start with 1"

let validate_length digits =
  match String.length digits with
  | 10 ->
    Ok digits
  | 11 ->
    validate_eleven_digit_number digits
  | length when length < 10 ->
    Error "must not be fewer than 10 digits"
  | _ ->
    Error "must not be greater than 11 digits"

(* ---- Area/Exchange code check ---- *)
let substring_of_digits digits position =
  String.sub digits ~pos:position ~len:3

let code_error code_type code_name =
  Error (code_type ^ " code cannot start with " ^ code_name)

let validate_area_and_exchange_code digits =
  let area_code = substring_of_digits digits 0 in
  let area_code_start = String.get area_code 0 in

  let exchange_code = substring_of_digits digits 3 in
  let exchange_code_start = String.get exchange_code 0 in

  match area_code_start, exchange_code_start with
  | '0', _ -> code_error "area" "zero"
  | '1', _ -> code_error "area" "one"
  | _, '0' -> code_error "exchange" "zero"
  | _, '1' -> code_error "exchange" "one"
  | _ -> Ok digits

(* ---- Main function ---- *)
let clean_and_validate_digits input =
  let digits_only = extract_digits input in
  match validate_length digits_only with
  | Error e ->
    Error e
  | Ok cleaned_digits ->
    validate_area_and_exchange_code cleaned_digits

let number input =
  if contains_letters input
    then Error "letters not permitted"
    else if contains_invalid_punctuation input
        then Error "punctuations not permitted"
        else clean_and_validate_digits input

File: solutions/python/phone-number/1/phone_number.py
"""Provide functionality for validating and formatting North American phone numbers."""

import re

COUNTRY_CODE = "1"
EXPECTED_DIGIT_COUNT = 10
MAXIMUM_DIGIT_COUNT = 11
AREA_CODE_LENGTH = 3
EXCHANGE_CODE_START = 3
EXCHANGE_CODE_END = 6
SUBSCRIBER_START = 6

VALID_PUNCTUATION = " -().+"

NUMBER_PATTERN = re.compile(r"^[0-9]+$")


class PhoneNumber:  # pylint: disable=too-few-public-methods
    """Represent a validated North American phone number."""

    def __init__(self, number: str):
        self.number: str = self._clean(number)

    @property
    def area_code(self) -> str:
        return self.number[:AREA_CODE_LENGTH]

    def pretty(self) -> str:
        area_code = self.number[:AREA_CODE_LENGTH]
        exchange_code = self.number[EXCHANGE_CODE_START:EXCHANGE_CODE_END]
        subscriber_number = self.number[SUBSCRIBER_START:]
        return f"({area_code})-{exchange_code}-{subscriber_number}"

    @staticmethod
    def _clean(number: str) -> str:
        if PhoneNumber._contains_letters(number):
            raise ValueError("letters not permitted")

        if PhoneNumber._contains_invalid_characters(number):
            raise ValueError("punctuations not permitted")

        digits = re.sub(r"\D", "", number)

        if len(digits) < EXPECTED_DIGIT_COUNT:
            raise ValueError("must not be fewer than 10 digits")

        if len(digits) > MAXIMUM_DIGIT_COUNT:
            raise ValueError("must not be greater than 11 digits")

        if len(digits) == MAXIMUM_DIGIT_COUNT:
            if not digits.startswith(COUNTRY_CODE):
                raise ValueError("11 digits must start with 1")

            digits = digits[1:]

        if digits[0] == "0":
            raise ValueError("area code cannot start with zero")

        if digits[0] == "1":
            raise ValueError("area code cannot start with one")

        if digits[EXCHANGE_CODE_START] == "0":
            raise ValueError("exchange code cannot start with zero")

        if digits[EXCHANGE_CODE_START] == "1":
            raise ValueError("exchange code cannot start with one")

        return digits

    @staticmethod
    def _contains_letters(number: str) -> bool:
        return any(
            character.isalpha()
            for character in number
        )

    @staticmethod
    def _contains_invalid_characters(number: str) -> bool:
        return any(
            not character.isdigit() and character not in VALID_PUNCTUATION
            for character in number
        )
